lib/wash/exe/wash.js

```diff
@@ -561,62 +555,84 @@ Wash.prototype.dispatch = function(pathSpec, argv, redirPathSpec, redirMode) {
-
-          var stdio = new NestedStdio(this.executeContext.stdio);
-          var redirBuffer;
-
-          var cxPromises = [
-            this.fileSystemManager.createExecuteContext(
-                result.path, stdio, arg)
-          ];
-
-          if (redirPathSpec) {
-            redirBuffer = new MemoryStreamBuffer();
-            stdio.stdout = redirBuffer.writableStream;
-            cxPromises.push(
-                this.fileSystemManager.createOpenContext(
-                    this.absPath(redirPathSpec), redirMode));
-          }
+      return this.findExecutable(pathSpec).then(function(result) {
+        return Promise.resolve({
+          fileSystem: this.fileSystemManager,
+          statResult: result.statResult,
+          path: result.path
+        });
+      }.bind(this));
+    }
+  }.bind(this));
+};
 
-          return Promise.all(cxPromises).then(
-            function(cxs) {
-              var execCx = cxs[0];
-              var redirCx = cxs[1];
-
-              execCx.onClose.addListener(function() {
-                stdio.close();
-              });
-
-              if (!!redirCx) {
-                redirCx.open().then(function() {
-                  redirBuffer.onData.addListener(function(data) {
-                    redirCx.write(0, SeekWhence.End, DataType.UTF8String, data);
-                  });
-                  redirBuffer.resume();
-                });
-              }
-              return this.dispatchExecuteContext(execCx);
-            }.bind(this));
-        }.bind(this));
+/**
+ * Run the given path with the given argv string, returning a promise that
+ * resolves to the result of the evaluation.
+ *
+ * For relative paths we'll search the builtins as well as $PATH.  The argv
+ * string will be parsed according to the sigil of the target executable.
+ *
+ * @param {string} pathSpec
+ * @param {string} argv
+ * @param {string} redirPathSpec
+ * @param {string} redirMode
+ * @return {!Promise<*>}
+ */
+Wash.prototype.dispatch = function(pathSpec, argv, redirPathSpec, redirMode) {
+  return this.findBuiltinOrExecutable(pathSpec).then(function(result) {
+    var arg = this.parseArgv(result.statResult.signature, argv);
+
+    var stdio = new NestedStdio(this.executeContext.stdio);
+    var redirBuffer;
+
+    var cxPromises = [
+      result.fileSystem.createExecuteContext(result.path, stdio, arg)
+    ];
+
+    if (redirPathSpec) {
+      redirBuffer = new MemoryStreamBuffer();
+      stdio.stdout = redirBuffer.writableStream;
+      cxPromises.push(
+          this.fileSystemManager.createOpenContext(
+              this.absPath(redirPathSpec), redirMode));
     }
+
+    return Promise.all(cxPromises).then(
+      function(cxs) {
+        var execCx = cxs[0];
+        var redirCx = cxs[1];
+
+        execCx.onClose.addListener(function(reason, value) {
+          if (reason === 'ok') {
+            if (typeof value != 'undefined' && typeof value != 'number' &&
```

Comment:
Although skipped printing for numbers seems strange: isn't it possible that an executable would legitimately return a number as a result?